omniledger/service/viewchange.go

```diff
@@ -0,0 +1,371 @@
+	"github.com/dedis/onet"
+	"github.com/dedis/onet/log"
+	"github.com/dedis/onet/network"
+	"github.com/dedis/protobuf"
+)
+
+type viewChangeManager struct {
+	sync.Mutex
+	controllers map[string]*viewchange.Controller
+}
+
+func newViewChangeManager() viewChangeManager {
+	return viewChangeManager{
+		controllers: make(map[string]*viewchange.Controller),
+	}
+}
+
+func (m *viewChangeManager) add(sendVC viewchange.SendInitReqFunc,
+	sendNV viewchange.SendNewViewReqFunc, isLeader viewchange.IsLeaderFunc, k string) {
+	m.Lock()
+	defer m.Unlock()
+	c := viewchange.NewController(sendVC, sendNV, isLeader)
+	m.controllers[k] = &c
+}
+
+func (m *viewChangeManager) start(myID network.ServerIdentityID, scID skipchain.SkipBlockID, initialDuration time.Duration, f int, k string) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[k]
+	go c.Start(myID, scID, initialDuration, f)
+}
+
+func (m *viewChangeManager) addReq(req viewchange.InitReq) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(req.View.Gen)]
+	c.AddReq(req)
+}
+
+func (m *viewChangeManager) addAnomaly(req viewchange.InitReq) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(req.View.Gen)]
+	c.AddAnomaly(req)
+}
+
+func (m *viewChangeManager) done(view viewchange.View) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(view.Gen)]
+	c.Done(view)
+	log.Lvl3("view-change done for " + view.String())
+}
+
+func (m *viewChangeManager) waiting(k string) bool {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[k]
+	return c.Waiting()
+}
+
+func (m *viewChangeManager) closeAll() {
+	m.Lock()
+	defer m.Unlock()
+	for _, c := range m.controllers {
+		c.Stop()
+	}
+	m.controllers = make(map[string]*viewchange.Controller)
+}
+
+// sendViewChangeReq is called when the node detects that a view change is
+// needed. It uses SendRaw to send the message to all other nodes. This
+// function should only be used as a callback in ViewChangeLog.
+func (s *Service) sendViewChangeReq(view viewchange.View) error {
+	log.Lvl2(s.ServerIdentity(), "sending view-change request for view:", view)
+	latest, err := s.db().GetLatestByID(view.ID)
+	if err != nil {
+		return err
+	}
+	req := viewchange.InitReq{
+		SignerID: s.ServerIdentity().ID,
+		View:     view,
+	}
+	if err := req.Sign(s.getPrivateKey()); err != nil {
+		return err
+	}
+	for _, sid := range latest.Roster.List {
+		if sid.Equal(s.ServerIdentity()) {
+			continue
+		}
+		if err := s.SendRaw(sid, &req); err != nil {
+			// Having an error here is fine because not all the
+			// nodes are guaranteed to be online. So we log a
+			// warning instead of returning an error.
+			log.Warn(s.ServerIdentity(), err)
+		}
+	}
+	return nil
+}
+
+func (s *Service) sendNewView(proof []viewchange.InitReq) {
+
+	if len(proof) == 0 {
+		log.Error(s.ServerIdentity(), "not enough proofs")
+	}
+	log.Lvl2(s.ServerIdentity(), "sending new-view request for view:", proof[0].View)
+
+	// Our own proof might not be signed, so sign it.
```

Comment:
Yes, so this happens when `viewchange.Controller` sends something to the anomaly channel by itself, because it didn't receive new-view message. The controller cannot sign because it's not a part of the service. A different way of doing it would be to register a "signRequest" callback.